src/core-ownership.js

```diff
@@ -1,13 +1,83 @@
-import { verifySignature } from '@mapeo/crypto'
+import { verifySignature, sign } from '@mapeo/crypto'
 import { NAMESPACES } from './core-manager/index.js'
 import { parseVersionId } from '@mapeo/schema'
 import assert from 'node:assert'
 import sodium from 'sodium-universal'
+import { kTable, kSelect, kCreateWithDocId } from './datatype/index.js'
+import { eq, or } from 'drizzle-orm'
+import mapObject from 'map-obj'
 
 /**
- * @typedef {Extract<ReturnType<import('@mapeo/schema').decode>, { schemaName: 'coreOwnership' }>} CoreOwnershipWithSignatures
+ * @typedef {import('./types.js').CoreOwnershipWithSignatures} CoreOwnershipWithSignatures
  */
 
+export class CoreOwnership {
+  #dataType
+  /**
+   *
+   * @param {object} opts
+   * @param {import('./datatype/index.js').DataType<import('./datastore/index.js').DataStore<'auth'>, typeof import('./schema/project.js').coreOwnershipTable, 'coreOwnership', import('@mapeo/schema').CoreOwnership, import('@mapeo/schema').CoreOwnershipValue>} opts.dataType
+   */
+  constructor({ dataType }) {
+    this.#dataType = dataType
+  }
+
+  /**
+   * @param {string} coreId
+   * @returns {Promise<string>} deviceId of device that owns the core
+   */
+  async getOwner(coreId) {
+    const table = this.#dataType[kTable]
+    const expressions = []
+    for (const namespace of NAMESPACES) {
+      expressions.push(eq(table[`${namespace}CoreId`], coreId))
+    }
+    // prettier-ignore
+    const result = this.#dataType[kSelect]()
+      .where(or.apply(null, expressions))
+      .get()
+    if (!result) {
+      throw new Error('NotFound')
+    }
+    return result.docId
+  }
+
+  /**
+   *
+   * @param {string} deviceId
+   * @param {typeof NAMESPACES[number]} namespace
+   * @returns {Promise<string>} coreId of core belonging to `deviceId` for `namespace`
+   */
+  async getCoreId(deviceId, namespace) {
+    const result = await this.#dataType.getByDocId(deviceId)
+    return result[`${namespace}CoreId`]
+  }
+
+  /**
+   *
+   * @param {import('./types.js').KeyPair} identityKeypair
+   * @param {Record<typeof NAMESPACES[number], import('./types.js').KeyPair>} coreKeypairs
+   */
+  async writeOwnership(identityKeypair, coreKeypairs) {
+    /** @type {import('./types.js').CoreOwnershipWithSignaturesValue} */
+    const docValue = {
+      schemaName: 'coreOwnership',
+      ...mapObject(coreKeypairs, (key, value) => {
```

Comment:
no suggestion here but was interested writing a limited version of this in case it would be useful to vendor since we're not using any of `map-obj`'s more advanced options in this case.

Seems to work (enough 😄 )

```ts
/**
 * @template {string} SourceKey
 * @template {Record<SourceKey, any>} Source
 * @template {string} TargetKey
 * @template TargetValue
 *
 * @param {Source} source
 * @param {(key: keyof Source, value: Source[keyof Source]) => [TargetKey, TargetValue]} mapper
 *
 * @returns {Record<TargetKey, TargetValue>}
 */
function mapObj(source, mapper) {
  const result = /** @type {Record<TargetKey, TargetValue>} */ ({})

  for (const [key, value] of Object.entries(source)) {
    const [newKey, newValue] = mapper(/** @type {keyof Source} */ (key), value)
    result[newKey] = newValue
  }

  return result
}
```

Reply:
It's a lot to vendor - would need to vendor all the type tests as well as the regular tests. I think sindre's modules are fairly reliable, I think it's ok to add as a dependency. I recently discovered `map-obj` and there are places where it would be better to use than what I have written in the past, and be more type-safe.